src/Onnx/OnnxToFeld.hs

```diff
@@ -116,38 +145,130 @@ buildValues td =
             , "  where "
-              <> L.intercalate "\n        " ("-- Nodes" : D.concatMap mkNode (G.node gr))
+              <> L.intercalate "\n        " (accesses ++ "-- Nodes" : D.concatMap (mkNode multiUses) (G.node gr))
             ]
   where name = mangle $ fromJust $ G.name gr
-        wName = "weights"
-        initVs = map (fromJust . TP.name) $ D.toList $ G.initializer gr
-        initSet = S.fromList initVs
-        params = wName : map mkParam inps
-        mkParam p = mangle $ vipName p
+        initSet = S.fromList $ map tiName $ concat initGroups
+        params = "(weights :: WeightRec)" : map (mangle . vipName) inps
         inps = filter (\ p -> vipName p `S.notMember` initSet) $ D.toList $ G.input gr
+        accesses = map mkAccess $ concat initGroups
+
+-- | Compute a string representation of the Haskell type of a group of initialized tensors
+--   with the same dimensionality and element type.
+mkITH :: TensorInfo -> L.ByteString
+mkITH ti = "Pull DIM1 (DPull (" <> mkSh (D.toList $ tiDims ti) <> ") " <> eTyH <>")"
+  where mkSh ds = "Z :. " <> L.intercalate " :. " ["Data Length" | _ <- ds]
+        eTyH = showElemT $ tiType ti
 
 -- | Construct a Feldspar pattern binding for an ONNX graph node
-mkNode :: N.NodeProto -> [L.ByteString]
-mkNode n = [outs <> " = " <> op <> " " <> attrs]
-         ++ [ "    " <> mangle v | v <- D.toList $ N.input n]
-         ++ [""]
+mkNode :: S.Set Utf8 -> N.NodeProto -> [L.ByteString]
+mkNode mUses n = [outs <> " = " <> forcing <> op <> " " <> attrs]
+              ++ [ "    " <> mangle v | v <- D.toList $ N.input n]
+              ++ [""]
   where outs = tuplify $ map mangle $ D.toList $ N.output n
         op = "onnx" <> strFromJ (N.op_type n) <> "_" <> show (D.length $ N.input n)
         attrs = "[" <> L.intercalate ", " (map showAttribute $ D.toList $ N.attribute n) <> "]"
+        forcing = if any (flip S.member mUses) $ N.output n then "force $ " else ""
 
--- | Mngle an ONNX node mane to a Feldspar identifier
+-- | Read from the weight record
+mkAccess :: TensorInfo -> L.ByteString
+mkAccess ti = vname <> " = sel (Proxy @" <> show (tiField ti) <> ") weights ! (Z :. " <> show (tiIdx ti) <> ")"
+  where vname = mangle $ tiName ti
+
+-- | Constuct the C code to read initialized tensors
+mkInitReadFile :: FilePath -> [[TensorInfo]] -> L.ByteString
+mkInitReadFile hf initGroups = L.concat $ start : map initVec initGroups ++ map initTensor (concat initGroups) ++ [end]
+  where start = L.unlines
+                [ "#include \"" <> fromString hf <> "\""
+                , ""
+                , "#include <stdlib.h>"
+                , "#include <stdio.h>"
+                , "#include <inttypes.h>"
+                , ""
+                , weightRecTC <> " * "
+                , "read_constants( char* wfile_name )"
+                , "{"
+                , "  FILE* wfile = fopen( wfile_name, \"r\" );"
```

Comment:
I think lost data files is a likely error scenario and there is no check of the return value from fopen.